### PR TITLE
feat: serde helpers for hashmaps and btreemaps with quantity key types

### LINKED ISSUE #1502: [Feature] Add serde helpers for handling maps with quantity keys
### Component

serde

### Describe the feature you would like

we already have helpers for handling vec

https://github.com/alloy-rs/alloy/blob/8857646d0851ab49252ed4478f8264b63804c63d/crates/serde/src/quantity.rs#L61-L64

it can be useful to have the same for hashmaps/btreemaps with quantity key types

## TODO
* add mods with manual map serde; take over this wip https://github.com/alloy-rs/alloy/pull/1503

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by stevencartavia**
Resolves #1502 

## Motivation
Existing `mod vec` implementation

## Solution
Created the same for hashmaps/btreemaps with quantity key types and tests.

## PR Checklist

- [x] Added Tests
- [ ] Added Documentation
- [ ] Breaking changes

### CODE REVIEW

File: crates/serde/src/quantity.rs [hunk 1]
```diff
@@ -164,6 +164,133 @@ pub mod u128_vec_vec_opt {
     }
 }
 
+/// Serde functions for encoding a hashmap of primitive numbers using the Ethereum "quantity"
```
Comment (onbjerg) [inline crates/serde/src/quantity.rs:167]:
```suggestion
/// Serde functions for encoding a `HashMap` of primitive numbers using the Ethereum "quantity"
```

File: crates/serde/src/quantity.rs [hunk 2]
```diff
@@ -164,6 +164,133 @@ pub mod u128_vec_vec_opt {
...
+/// format.
+///
+/// See [`quantity`](self) for more information.
+pub mod hashmap {
+    use super::private::ConvertRuint;
+    use core::{fmt, marker::PhantomData};
+    use serde::{de::MapAccess, ser::SerializeMap, Deserializer, Serializer};
+    use std::collections::HashMap;
+
+    /// Serializes a hashmap of primitive numbers as a "quantity" hex string.
```
Comment (onbjerg) [inline crates/serde/src/quantity.rs:177]:
```suggestion
    /// Serializes a `HashMap` of primitive numbers as a "quantity" hex string.
```

File: crates/serde/src/quantity.rs [hunk 3]
```diff
@@ -164,6 +164,133 @@ pub mod u128_vec_vec_opt {
...
+        S: Serializer,
+    {
+        let mut map_ser = serializer.serialize_map(Some(map.len()))?;
+        for (key, value) in map {
+            map_ser.serialize_entry(&key.into_ruint(), &value.into_ruint())?;
+        }
+        map_ser.end()
+    }
+
+    /// Deserializes a hashmap of primitive numbers from a "quantity" hex string.
```
Comment (onbjerg) [inline crates/serde/src/quantity.rs:191]:
```suggestion
    /// Deserializes a `HashMap` of primitive numbers from a "quantity" hex string.
```

File: crates/serde/src/quantity.rs [hunk 4]
```diff
@@ -164,6 +164,133 @@ pub mod u128_vec_vec_opt {
...
+                Ok(values)
+            }
+        }
+
+        let visitor = HashMapVisitor { marker: PhantomData };
+        deserializer.deserialize_map(visitor)
+    }
+}
+
+/// Serde functions for encoding a BTreeMap of primitive numbers using the Ethereum "quantity"
```
Comment (onbjerg) [inline crates/serde/src/quantity.rs:231]:
```suggestion
/// Serde functions for encoding a `BTreeMap` of primitive numbers using the Ethereum "quantity"
```

**Review by onbjerg — CHANGES_REQUESTED** [review]
some comment nits,

this is also not quite right: we are only supposed to encode the keys as quantity, the value can be anything

File: crates/serde/src/quantity.rs [hunk 5]
```diff
@@ -164,6 +164,133 @@ pub mod u128_vec_vec_opt {
...
+
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                formatter.write_str("a map with quantity hex-encoded keys and values")
+            }
+
+            fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
+            where
+                A: MapAccess<'de>,
+            {
+                let mut values = HashMap::new();
```
Comment (mattsse) [inline crates/serde/src/quantity.rs:217]: this can pre allocate with the map's size hint

File: crates/serde/src/quantity.rs [hunk 6]
```diff
@@ -164,6 +164,133 @@ pub mod u128_vec_vec_opt {
...
+///
+/// See [`quantity`](self) for more information.
+pub mod hashmap {
+    use super::private::ConvertRuint;
+    use core::{fmt, marker::PhantomData};
+    use serde::{de::MapAccess, ser::SerializeMap, Deserializer, Serializer};
+    use std::collections::HashMap;
+
+    /// Serializes a hashmap of primitive numbers as a "quantity" hex string.
+    pub fn serialize<K, V, S>(map: &HashMap<K, V>, serializer: S) -> Result<S::Ok, S::Error>
```
Comment (mattsse) [inline crates/serde/src/quantity.rs:178]: this should also relax the hasher generic, see for ref https://github.com/serde-rs/serde/blob/418062165f9fe395461db9f61569c3142c584854/serde/src/ser/impls.rs#L495-L506

File: crates/serde/src/quantity.rs [hunk 7]
```diff
@@ -164,6 +164,133 @@ pub mod u128_vec_vec_opt {
...
+    {
+        let mut map_ser = serializer.serialize_map(Some(map.len()))?;
+        for (key, value) in map {
+            map_ser.serialize_entry(&key.into_ruint(), &value.into_ruint())?;
+        }
+        map_ser.end()
+    }
+
+    /// Deserializes a hashmap of primitive numbers from a "quantity" hex string.
+    pub fn deserialize<'de, K, V, D>(deserializer: D) -> Result<HashMap<K, V>, D::Error>
```
Comment (mattsse) [inline crates/serde/src/quantity.rs:192]: this should also relax the hasher

**Review by mattsse — CHANGES_REQUESTED** [review]

**Comment by stevencartavia** [comment]
@onbjerg @mattsse, done with the requested changes! :)

**Review by mattsse — APPROVED** [review]

**Pull request merged**